### DISCUSSION TITLE
Fixed constraint on cable nodes

### DISCUSSION THREAD

**Discussion by Fokion-Sanoudos** [SofaDefrost/SoftRobots#140] [Help - Programming with SoftRobots] [answered]
I am trying to use the cable constraint of SoftRobots as beam in order to push a surface. The problem is that when I try to extend the beam the node of the beam moves and rotates instead of being fixed. I don't want to use a pull point because I want to the whole structure to move but I was wondering if fixed rotation or general constraints can be added to the cable node.

Edit: I have uploaded a zip file with the simulation files.
[Simulation.zip](https://github.com/SofaDefrost/SoftRobots/files/8639801/Simulation.zip)

**Comment by EulalieCoevoet** [comment]
There is a component in SOFA called `PartialFixedConstraint` that you can use in addition to the `CableConstraint` to do that I think. 
With this component you can constrain some points of your model to be fixed in chosen directions. 
I don't think adding this component to the cable node will work though, you should probably add it directly to the "surface" node.

**Reply by Fokion-Sanoudos** [reply]
Thank you for the reply. I am aware of the component you mentioned; however, the nodes of the cable cannot be constrained with it. As in, 
PartialFixedConstraint needs to address some mesh indices, but the nodes of the CableConstraint cannot be added to a mechanical object to be labeled as indices. At least, that's what I found so far.

**Reply by EulalieCoevoet** [reply]
I imagine your code looking something like that:

```python
surface = rootnode.addChild('Surface')
surface.addObject('YOUR_SOLVERS_MESHLOADER_TOPOLOGY')
surface.addObject('MechanicalObject')
# You can use an engine to select the indices of your model to be partially fixed (BoxROI for instance)
surface.addObject('YOUR_ROI_ENGINE', name='roi')  
# Use this component to fix the selected points in all directions except x for instance
surface.addObject('PartialFixedConstraint', indices='@roi.indices', fixedDirections=[0,1,1])  
surface.addObject('YOUR_MASS_AND_FORCEFIELD')

# You can either directly pull/push a point of your FEM mesh that as been partially fixed
cable.addObject('CableConstraint', indices=[YOUR_INDEX])

# OR pull/push a point that is located inside an FEM element (tetrahedron I guess) that has been partially fixed 
# and use a mapping
cable = surface.addChild('Cable') 
cable.addObject('MechanicalObject', position=[YOUR_POSITION])
cable.addObject('CableConstraint', indices=[0])
cable.addObject('BarycentricMapping')
```

**Reply by EulalieCoevoet** [reply]
@Fokion-Sanoudos does it work with this idea?

**Reply by Fokion-Sanoudos** [reply]
I have already been doing that with my code. It kinda works but I was looking to constraint the rotation or generally constraint somehow the movement of the "pull point" of the cable. A solution I found is that I constrained the rest of the body in one direction to stop the rotations. It is not ideal but it will have to do.

**Reply by EulalieCoevoet** [reply]
Hi @Fokion-Sanoudos, could you write a comment with your final solution and mark it as the answer?

**Reply by Fokion-Sanoudos** [reply]
It's really not the answer to the question it is just a workaround that I implemented.

**Comment by Fokion-Sanoudos** [comment, marked as answer]
This is what I did at the end:
`ring.addObject('BoxROI', name='roic', box=[15, 81/2, -420/2, -15, -81/2, 420/2], drawBoxes=True)
ring.addObject('PartialFixedConstraint', fixedDirections= [1, 1, 0],indices=ring.roic.indices.getLinkPath())`
